### PR TITLE
Add genre derived query

### LINKED ISSUE #11: Add derived query to find books by genre
Add a derived query to find by book genre (using our enum). This will also need to be added to the query resolver.

Accompanying tests will be required.

### PR DESCRIPTION
**Pull request by shivansh29**
## Summary of change

Derived query by genre 

## Related issue

Closes #11 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-jcijyenp-JiKFGBv62FIPoFnvOW6Ubg)

### CODE REVIEW

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java [hunk 1]
```diff
@@ -19,10 +19,13 @@
...
 import org.springframework.data.repository.CrudRepository;
+import org.springframework.data.repository.query.Param;
 
 public interface BookRepository extends CrudRepository<Book, Long> {
   
     @Query("SELECT DISTINCT b FROM Book b LEFT JOIN FETCH b.authors")
     List<Book> findAllBooks();
   
+    @Query("FROM Book where title = :booktitle")
+    Book findByTitle(@Param("booktitle") String title);
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/repository/BookRepository.java:30]: Could you please change this to not use JPQL? It should be

```java
Book findByTitle(String bookTitle);
```

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this. Please address the review comment

**Comment by knjk04** [comment]
Also, please resolve the merge conflict

**Comment by knjk04** [comment]
How are you getting on with this?

**Comment by shivansh29** [comment]
Will Complete it in 2 days

**Comment by knjk04** [comment]
I noticed you committed. Is this ready for another review?

**Comment by shivansh29** [comment]
Yes

File: src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java [hunk 2]
```diff
@@ -40,7 +40,12 @@ public QueryResolver(AuthorRepository authorRepository, BookRepository bookRepos
         return authorRepository.findAllAuthors();
     }
 
+    
+    public Book findByTitle(String title) {
+    	return bookRepository.findByBooktitle(title);
+
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java:46]: You're missing a closing brace here

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java [hunk 3]
```diff
@@ -19,11 +19,15 @@
...
 import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.CrudRepository;
+import org.springframework.data.repository.query.Param;
 
 public interface BookRepository extends CrudRepository<Book, Long> {
   
     @Query("SELECT DISTINCT b FROM Book b LEFT JOIN FETCH b.authors")
     List<Book> findAllBooks();
 
+    Book findByBooktitle( String booktitle);
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/repository/BookRepository.java:29]: This issue was meant to create a derived query to find books by genre, not title. Could you please change this?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this.

Please see my review comments. Also, this doesn't compile for me (even after adding in the missing closing brace). One of the reasons is because a derived query cannot be made. The error message is quite descriptive, so I'll let you figure it out. If you need help, however, feel free to create a new discussion on GitHub or join our Slack channel.

**Comment by knjk04** [comment]
How are you getting on with this?

**Comment by shivansh29** [comment]
Please check I have merged it with new Changes

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java [hunk 4]
```diff
@@ -27,5 +28,7 @@
 
     Book findBookByIsbn13(String isbn13);
   
+    Book findByGenre( String genre);
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/repository/BookRepository.java:31]: There may be multiple books in a genre, so could you please change the return type to a `List<Book>`

File: src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java (hunk 2 above)
Comment (knjk04) [reply src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java:46]: This is still missing

File: src/main/java/com/karankumar/booksapi/repository/BookRepository.java (hunk 4 above)
Comment (knjk04) [reply src/main/java/com/karankumar/booksapi/repository/BookRepository.java:31]: Also, we don't want to search by a `String`. Rather, we would like to pass in our enum, `BookGenre`.

File: src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java [hunk 5]
```diff
@@ -39,12 +39,17 @@ public QueryResolver(AuthorRepository authorRepository, BookRepository bookRepos
     public Iterable<Author> findAllAuthors() {
         return authorRepository.findAllAuthors();
     }
+    
+    public Book findByGenre(String genre) {
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java:43]: Change argument type to our genre enum

File: src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java [hunk 6]
```diff
@@ -39,12 +39,17 @@ public QueryResolver(AuthorRepository authorRepository, BookRepository bookRepos
...
+    	return bookRepository.findByGenre(genre);
+
 
     public Book findBookByIsbn13(String isbn13) {
         return bookRepository.findBookByIsbn13(isbn13);
     }
 
      public Book findByTitleIgnoreCase(String title) {
         return bookRepository.findByTitleIgnoreCase(title);
+
```
Comment (knjk04) [inline src/main/java/com/karankumar/booksapi/resolver/QueryResolver.java:53]: Could you please remove this extraneous line?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks. I've left some feedback as this is not quite right yet. Please ensure you get it to compile and work correctly (the missing closing brace was not still added).

This is not showing up on `localhost:8080/playground`. You will need to also add the query signature to our schema file in `src/main/resources/graphql/query.graphqls`. Please use the new signature where the argument is of type `Genre`, which corresponds to our enum, `genre.graphqls`.

> Ensure your commit messages are descriptive (i.e. a commit message should effectively summarise what change(s) you made)

The above is from our contributing document. Please note that your current commit messages do not summarise what you did. Could you please fix this (e.g. with an interactive rebase)?

**Comment by shivansh29** [comment]
I'll  close this pull request  and create a new one    
I hope that's fine by you

**Comment by knjk04** [comment]
That's fine

**Pull request closed without merging**